hypervisor/arch/x86/vtd.c

```diff
@@ -833,112 +850,111 @@ static int32_t add_iommu_device(struct iommu_domain *domain, uint16_t segment, u
 	uint64_t upper;
 	uint64_t lower = 0UL;
 	struct acrn_vm *vm;
+	int32_t ret = 0;
 
 	dmar_unit = device_to_dmaru(segment, bus, devfun);
 	if (dmar_unit == NULL) {
 		pr_err("no dmar unit found for device: %x:%x.%x", bus, pci_slot(devfun), pci_func(devfun));
-		return -EINVAL;
-	}
-
-	if (dmar_unit->drhd->ignore) {
+		ret = -EINVAL;
+	} else if (dmar_unit->drhd->ignore) {
 		dev_dbg(ACRN_DBG_IOMMU, "device is ignored :0x%x:%x.%x", bus, pci_slot(devfun), pci_func(devfun));
-		return 0;
-	}
-
-	if (!dmar_unit_support_aw(dmar_unit, domain->addr_width)) {
+	} else if (!dmar_unit_support_aw(dmar_unit, domain->addr_width)) {
 		pr_err("dmar doesn't support addr width %d", domain->addr_width);
-		return -EINVAL;
-	}
-
-	if (iommu_ecap_sc(dmar_unit->ecap) == 0U) {
-		vm = get_vm_from_vmid(domain->vm_id);
-		if (vm != NULL) {
-			vm->snoopy_mem = false;
+		ret = -EINVAL;
+	} else if (dmar_unit->root_table_addr == 0UL) {
+		pr_err("dmar root table address invalid");
```

Comment:
This else if can be merged with previous one...

Reply:
But they lead to different error message, sorry if i misunderstood...

Reply:
you can use a common error message

Reply:
Done.